File: src/tasks/cli/handlers/AddHandler.java
package tasks.cli.handlers;

import static java.util.Objects.requireNonNull;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;
import omnia.data.cache.Memoized;
import omnia.data.structure.Set;
import omnia.data.structure.immutable.ImmutableSet;
import tasks.cli.arg.AddArguments;
import tasks.model.Task;
import tasks.model.TaskBuilder;
import tasks.model.TaskStore;

public final class AddHandler implements ArgumentHandler<AddArguments> {
  private final Memoized<TaskStore> taskStore;

  AddHandler(Memoized<TaskStore> taskStore) {
    this.taskStore = requireNonNull(taskStore);
  }

  @Override
  public Completable handle(AddArguments arguments) {
    // Validate arguments
    String label = arguments.description().trim();
    if (label.isEmpty()) {
      throw new HandlerException("description cannot be empty or whitespace only");
    }

    // Collect the dependencies and dependents
    Set<Task> blockingTasks = ImmutableSet.copyOf(arguments.blockingTasks());
    Set<Task> blockedTasks = ImmutableSet.copyOf(arguments.blockedTasks());

    TaskStore taskStore = this.taskStore.value();

    // Construct the new task, commit to disk, print output
    return taskStore.createTask(
        label,
        builder ->
            Single.just(builder)
                .flatMap(b ->
                    Observable.fromIterable(blockingTasks).reduce(b, TaskBuilder::addBlockingTask))
                .flatMap(b ->
                    Observable.fromIterable(blockedTasks).reduce(b, TaskBuilder::addBlockedTask))
                .blockingGet())
        .flatMap(task -> taskStore.writeToDisk().toSingleDefault(task))
        .map(task -> "task created: " + task)
        .flatMapCompletable(msg -> Completable.fromAction(() -> System.out.println(msg)));
  }
}

File: cli/src/me/strangepan/tasks/cli/feature/add/AddHandler.kt
package me.strangepan.tasks.cli.feature.add

import io.reactivex.rxjava3.core.Single
import omnia.cli.out.Output
import omnia.cli.out.Output.Companion.builder
import omnia.data.cache.Memoized
import me.strangepan.tasks.cli.command.common.CommonArguments
import me.strangepan.tasks.cli.handler.ArgumentHandler
import me.strangepan.tasks.cli.handler.HandlerException
import me.strangepan.tasks.cli.model.render
import me.strangepan.tasks.engine.model.ObservableTaskStore

/** Business logic for the Add command.  */
class AddHandler(private val taskStore: Memoized<out ObservableTaskStore>) : ArgumentHandler<AddArguments> {

  override fun handle(arguments: CommonArguments<out AddArguments>): Single<Output> {
    // Validate arguments
    val description = arguments.specificArguments().description().trim { it <= ' ' }
    if (description.isEmpty()) {
      throw HandlerException("description cannot be empty or whitespace only")
    }

    // Construct the new task, commit to disk, print output
    return taskStore.value()
        .createTask(description) {
            it.setBlockingTasks(arguments.specificArguments().blockingTasks())
                .setBlockedTasks(arguments.specificArguments().blockedTasks())
    }
      .map { it.third() }
      .map { it.render() }
      .map { builder().append("task created: ").append(it).build() }
      .map { builder().appendLine(it).build() }
  }

}